File: backend/app/services/deals.py
"""团购工坊业务逻辑 — 毛利估算（SPEC-DEALS §4 口径）。"""
from __future__ import annotations

from decimal import Decimal
from typing import Any

# 平台佣金行业均值默认值（MVP：不支持门店级实际费率覆盖，人工编辑可改）
# 口径与 SPEC-DEALS §4 / §5 一致，后续可改为配置项。
DEFAULT_COMMISSION_RATES: dict[str, float] = {
    "douyin": 0.06,
    "meituan": 0.08,
    "xiaohongshu": 0.05,
}


def platform_commission_rate(platform: str) -> float:
    """按项目主平台取默认佣金率，未知平台回落抖音默认值。"""
    return DEFAULT_COMMISSION_RATES.get(platform, DEFAULT_COMMISSION_RATES["douyin"])


def compute_margins(
    deal_price: Decimal | float,
    cost_estimate: Decimal | float,
    commission_rate: float,
) -> dict[str, float]:
    """按 SPEC 公式计算毛利：
    gross_margin = (deal_price - 组合成本) / deal_price
    net_margin = (deal_price × (1 - commission) - 组合成本) / deal_price
    """
    deal_f = float(deal_price)
    cost_f = float(cost_estimate)
    if deal_f <= 0:
        raise ValueError("deal_price 必须为正")
    gross = (deal_f - cost_f) / deal_f
    net = (deal_f * (1 - commission_rate) - cost_f) / deal_f
    return {
        "gross_margin": round(gross, 4),
        "platform_commission_rate": commission_rate,
        "net_margin": round(net, 4),
    }


def build_margin_estimate(
    deal_price: Decimal,
    cost_estimate: Decimal,
    commission_rate: float,
    *,
    estimated: bool = False,
    scheme_type: str = "",
) -> dict[str, Any]:
    """生成 margin_estimate 完整结构（含 note 警示）。"""
    margins = compute_margins(deal_price, cost_estimate, commission_rate)
    notes: list[str] = []
    if estimated:
        notes.append("含 AI 估算成本，请按实际成本校正")
    if scheme_type == "hook":
        notes.append("引流款核心是拉新到店，毛利允许偏低")
    if margins["net_margin"] < 0:
        notes.append("净毛利为负：请重新评估组合/定价，确认后再上线")
    return {
        **margins,
        "note": "；".join(notes),
    }

File: frontend/src/utils/deals.ts
import type { DealPlatform, DealScheme, DealSchemeCopy, MarginEstimate } from "../services/deals";

export const PLATFORMS: DealPlatform[] = ["douyin", "meituan", "xiaohongshu"];

export const PLATFORM_LABELS: Record<string, string> = {
  douyin: "抖音",
  meituan: "美团",
  xiaohongshu: "小红书",
};

export const CATEGORY_LABELS: Record<string, string> = {
  signature: "招牌",
  staple: "主食",
  snack: "小吃",
  drink: "饮品",
};

export const SCHEME_TYPE_LABELS: Record<string, string> = {
  hook: "引流款",
  profit: "利润款",
  scenario: "场景款",
};

export function percent(value: number | null | undefined): string {
  if (value === null || value === undefined || Number.isNaN(Number(value))) return "-";
  return `${(Number(value) * 100).toFixed(1)}%`;
}

export function formatPrice(value: string | number | null | undefined): string {
  if (value === null || value === undefined || value === "") return "-";
  const n = Number(value);
  return Number.isNaN(n) ? "-" : `¥${n.toFixed(2)}`;
}

export function marginLines(margin: MarginEstimate | null | undefined): {
  gross: string;
  net: string;
  rate: string;
  note: string;
} {
  if (!margin) return { gross: "-", net: "-", rate: "-", note: "" };
  return {
    gross: percent(margin.gross_margin),
    net: percent(margin.net_margin),
    rate: percent(margin.platform_commission_rate),
    note: margin.note || "",
  };
}

export function hasActiveSchemes(schemes: DealScheme[]): boolean {
  return schemes.some((s) => !s.is_archived);
}

export function hasEditedScheme(schemes: DealScheme[]): boolean {
  return schemes.some((s) => !s.is_archived && s.status === "edited");
}

export function shouldConfirmRegenerate(schemes: DealScheme[]): boolean {
  return hasActiveSchemes(schemes) || hasEditedScheme(schemes);
}

export function copyForPlatform(
  scheme: DealScheme,
  platform: string
): DealSchemeCopy | undefined {
  return (scheme.copies ?? []).find((c) => c.platform === platform);
}

export function schemeItemLines(scheme: DealScheme): string {
  return (scheme.items ?? [])
    .map((it) => `${it.name} ×${it.qty}`)
    .join(" + ");
}

export function schemeSummary(scheme: DealScheme): string {
  const lines = [
    `组合：${schemeItemLines(scheme) || "-"}`,
    `原价 ${formatPrice(scheme.original_price)} → 团购价 ${formatPrice(scheme.deal_price)}`,
  ];
  return lines.join("\n");
}

export function buildCopyClipboardText(copy: DealSchemeCopy | undefined): string {
  if (!copy) return "";
  const points = (copy.selling_points ?? []).map((p) => `- ${p}`).join("\n");
  return [copy.title, points, copy.rules ? `规则：${copy.rules}` : ""]
    .filter(Boolean)
    .join("\n\n");
}
